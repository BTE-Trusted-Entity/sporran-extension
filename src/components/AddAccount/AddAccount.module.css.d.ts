declare namespace AddAccountModuleCssNamespace {
  export interface IAddAccountModuleCss {
    toggle: string;
  }
}

declare const AddAccountModuleCssModule: AddAccountModuleCssNamespace.IAddAccountModuleCss & {
  /** WARNING: Only available when `css-loader` is used without `style-loader` or `mini-css-extract-plugin` */
  locals: AddAccountModuleCssNamespace.IAddAccountModuleCss;
};

export = AddAccountModuleCssModule;
