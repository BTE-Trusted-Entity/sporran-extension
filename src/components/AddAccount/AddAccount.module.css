.toggle {
  composes: toggle from '../../components/Menu/Menu.module.css';
  composes: buttonIcon from '../../components/Button/Button.module.css';
  width: 1.625rem;
  height: 1.625rem;
  background-image: url('../../components/images/plus-circle.svg');
}
