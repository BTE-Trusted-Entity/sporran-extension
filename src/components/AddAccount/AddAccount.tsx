import { useCallback } from 'react';
import useDropdownMenu from 'react-accessible-dropdown-menu-hook';
import { Link } from 'react-router-dom';
import { browser } from 'webextension-polyfill-ts';

import { paths } from '../../views/paths';
import { useAccounts } from '../../utilities/accounts/accounts';

import menuStyles from '../Menu/Menu.module.css';
import styles from './AddAccount.module.css';

export function AddAccount(): JSX.Element | null {
  const t = browser.i18n.getMessage;

  const accounts = useAccounts().data;
  const hasAccounts = accounts && Object.values(accounts).length > 0;

  const { buttonProps, itemProps, isOpen, setIsOpen } = useDropdownMenu(2);

  const handleClick = useCallback(() => {
    setIsOpen(false);
  }, [setIsOpen]);

  if (!hasAccounts) {
    return null;
  }

  return (
    <div className={menuStyles.wrapper}>
      <button
        {...buttonProps}
        type="button"
        className={styles.toggle}
        title={t('component_AddAccount_label')}
        aria-label={t('component_AddAccount_label')}
      />

      {isOpen && (
        <div className={menuStyles.dropdown} role="menu" onClick={handleClick}>
          <h4 className={menuStyles.heading}>
            {t('component_AddAccount_label')}
          </h4>

          <ul className={menuStyles.list}>
            <li className={menuStyles.listItem}>
              <Link {...itemProps[0]} to={paths.account.create.start}>
                {t('component_AddAccount_create')}
              </Link>
            </li>

            <li className={menuStyles.listItem}>
              <Link {...itemProps[1]} to={paths.account.import.start}>
                {t('component_AddAccount_import')}
              </Link>
            </li>
          </ul>
        </div>
      )}
    </div>
  );
}
