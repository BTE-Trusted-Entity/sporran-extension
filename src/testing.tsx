export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
