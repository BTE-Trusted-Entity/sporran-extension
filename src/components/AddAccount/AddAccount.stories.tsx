import { Meta } from '@storybook/react';

import { AddAccount } from './AddAccount';

export default {
  title: 'Components/AddAccount',
  component: AddAccount,
} as Meta;

function mockUseAccounts() {
  return {
    data: {
      '4tJbxxKqYRv3gDvY66BKyKzZheHEH8a27VBiMfeGX2iQrire': {
        name: 'My Sporran Account',
        address: '4tJbxxKqYRv3gDvY66BKyKzZheHEH8a27VBiMfeGX2iQrire',
        index: 1,
      },
    },
    revalidate: async () => true,
    mutate: () => undefined,
    isValidating: false,
  };
}

export function Template(): JSX.Element {
  return <AddAccount useAccounts={mockUseAccounts} />;
}
