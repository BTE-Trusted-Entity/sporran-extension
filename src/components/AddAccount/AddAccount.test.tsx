import userEvent from '@testing-library/user-event';

import { render, screen } from '../../testing';
import { AddAccount } from './AddAccount';

describe('AddAccount', () => {
  it('should render', async () => {
    const { container } = render(<AddAccount />);
    expect(container).toMatchSnapshot();
  });

  it('menu should be visible when menu button clicked', async () => {
    render(<AddAccount />);
    const menu = await screen.findByRole('menu');
    expect(menu).not.toHaveClass('visible');

    // Both tests fail. Why?
    expect(menu).not.toBeVisible();
    expect(menu).toHaveStyle('visibility: hidden');

    userEvent.click(await screen.findByRole('button', { name: '+' }));
    expect(menu).toHaveClass('visible');
    expect(menu).toBeVisible();
  });
});
