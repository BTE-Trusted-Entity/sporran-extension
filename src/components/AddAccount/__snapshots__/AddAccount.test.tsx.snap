// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddAccount menu should be visible when menu button clicked 1`] = `
<div>
  <nav
    class="container"
  >
    <button
      aria-expanded="true"
      aria-haspopup="true"
      aria-label="Add account"
      class="button"
      role="button"
      tabindex="0"
      title="Add account"
      type="button"
    >
      +
    </button>
    <div
      class="menu"
      role="menu"
    >
      <h4
        class="menuHeading"
      >
        Add account
      </h4>
      <ul
        class="list"
      >
        <li
          class="listItem"
        >
          <a
            href="/account/create"
            role="menuitem"
            tabindex="-1"
          >
            Create new account
          </a>
        </li>
        <li
          class="listItem"
        >
          <a
            href="/account/import"
            role="menuitem"
            tabindex="-1"
          >
            Import account from pre-existing phrase
          </a>
        </li>
      </ul>
    </div>
  </nav>
</div>
`;

exports[`AddAccount should render 1`] = `
<div>
  <nav
    class="container"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Add account"
      class="button"
      role="button"
      tabindex="0"
      title="Add account"
      type="button"
    >
      +
    </button>
  </nav>
</div>
`;
